From 50d6b5a9fdb6a62661075badae6646767756d88c Mon Sep 17 00:00:00 2001
From: Min Fanlei <minfanlei@wxiat.com>
Date: Wed, 26 Oct 2022 09:03:49 +0800
Subject: [PATCH] sw64: kvm: fix guest longtime offset of VCPU
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 50d6b5a9fdb6a62661075badae6646767756d88c
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I645O8

--------------------------------

Other VCPUs have to sync longtime offset with VCPU0 to ensure
the consistency of clock source, and this will avoid RCU CPU
stalls in guest live migration.

Signed-off-by: Min Fanlei <minfanlei@wxiat.com>
Reviewed-by: He Sheng <hesheng@wxiat.com>
Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/kvm/kvm-sw64.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/arch/sw_64/kvm/kvm-sw64.c b/arch/sw_64/kvm/kvm-sw64.c
index be5218234adf..c5a16d1c350b 100644
--- a/arch/sw_64/kvm/kvm-sw64.c
+++ b/arch/sw_64/kvm/kvm-sw64.c
@@ -23,6 +23,8 @@
 #include "vmem.c"
 
 bool set_msi_flag;
+static unsigned long longtime_offset;
+
 #if defined(CONFIG_DEBUG_FS) && defined(CONFIG_NUMA)
 extern bool bind_vcpu_enabled;
 #endif
@@ -726,10 +728,13 @@ long kvm_arch_vcpu_ioctl(struct file *filp,
 			vcpu->arch.vcb.vpcr
 				= get_vpcr(vcpu->kvm->arch.host_phys_addr, vcpu->kvm->arch.size, 0);
 
-			result = sw64_io_read(0, LONG_TIME);
-
 			/* synchronize the longtime of source and destination */
-			vcpu->arch.vcb.guest_longtime_offset = vcpu->arch.vcb.guest_longtime - result;
+			if (vcpu->arch.vcb.whami == 0) {
+				result = sw64_io_read(0, LONG_TIME);
+				vcpu->arch.vcb.guest_longtime_offset = vcpu->arch.vcb.guest_longtime - result;
+				longtime_offset = vcpu->arch.vcb.guest_longtime_offset;
+			} else
+				vcpu->arch.vcb.guest_longtime_offset = longtime_offset;
 
 			set_timer(vcpu, 200000000);
 			vcpu->arch.vcb.migration_mark = 0;
-- 
2.33.0

